src/azul/transformer.py

```diff
@@ -42,6 +48,91 @@ class Document:
     version_type: ClassVar[Optional[str]] = None
     contents: Optional[JSON]
 
+    @classmethod
+    @lru_cache(maxsize=1)
+    def field_types(cls):
+        from azul.plugin import Plugin
+        return Plugin.load().field_types()
+
+    @classmethod
+    @lru_cache(maxsize=64)
+    def field_type(cls, path):
+        """
+        Traverse cls.field_types nested dict with given path and return value
+        :param path: A tuple of keys to traverse down the field_types dict
+        """
+        types_mapping = cls.field_types()
+        for p in path:
+            try:
+                types_mapping = types_mapping[p]
+            except KeyError:
+                return None
+            except TypeError:
+                return None
+        return types_mapping
+
+
+    @classmethod
+    def translate_field(cls, value, path=(), forward=True):
+        """
+        Translate a single value for insert into or after fetching from Elasticsearch.
+
+        :param value: A value to translate
+        :param path: A tuple representing the field name split on dots (.)
+        :param forward: If we should translate forward or backward (aka un-translate)
+        :return: A copy of the original document with values translated
+        """
+        field_type = cls.field_type(path)
+        if field_type is bool:
+            if forward:
+                return {None: -1, False: 0, True: 1}[value]
+            else:
+                return {-1: None, 0: False, 1: True}[value]
+        elif field_type is int or field_type is float:
+            if forward:
+                if value is None:
+                    return MIN_INT
+            else:
+                if value == MIN_INT:
+                    return None
+            return value
+        elif field_type is str:
+            if forward:
+                if value is None:
+                    return config.null_keyword
+            else:
+                if value == config.null_keyword:
+                    return None
+            return value
+        elif field_type is dict:
+            return value
+        elif field_type is api.UUID4:
+            return value
+        elif field_type is None:
+            return value
+        else:
+            raise ValueError(f'Field type not found for path {path}')
+
+    @classmethod
+    def translate_fields(cls, contents, path=(), forward=True):
```

Comment:
```suggestion
    def translate_fields(cls, source, path=(), forward=True):
```

I'm worried that the term contents conflicts with the top level key of the same name inside the source JSON.

Reply:
Could you also add type annotations, `source: JSON` for example?